File: src/main/java/no/nav/sosialhjelp/soknad/consumer/virusscan/ClamAvVirusScanner.java
package no.nav.sosialhjelp.soknad.consumer.virusscan;

import io.github.resilience4j.retry.Retry;
import no.nav.sosialhjelp.soknad.domain.model.exception.OpplastingException;
import no.nav.sosialhjelp.soknad.domain.model.util.ServiceUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import org.springframework.web.client.HttpServerErrorException;
import org.springframework.web.reactive.function.BodyInserters;
import org.springframework.web.reactive.function.client.WebClient;

import java.net.URI;

import static no.nav.sosialhjelp.soknad.consumer.retry.RetryUtils.DEFAULT_EXPONENTIAL_BACKOFF_MULTIPLIER;
import static no.nav.sosialhjelp.soknad.consumer.retry.RetryUtils.DEFAULT_INITIAL_WAIT_INTERVAL_MILLIS;
import static no.nav.sosialhjelp.soknad.consumer.retry.RetryUtils.DEFAULT_MAX_ATTEMPTS;
import static no.nav.sosialhjelp.soknad.consumer.retry.RetryUtils.retryConfig;
import static no.nav.sosialhjelp.soknad.consumer.retry.RetryUtils.withRetry;
import static no.nav.sosialhjelp.soknad.consumer.virusscan.Result.OK;

@Component
public class ClamAvVirusScanner implements VirusScanner {

    private static final Logger logger = LoggerFactory.getLogger(ClamAvVirusScanner.class);

    private final WebClient webClient;
    private final Retry retry;

    @Value("${soknad.vedlegg.virusscan.enabled}")
    private boolean enabled;

    public ClamAvVirusScanner(URI uri, WebClient webClient) {
        this.webClient = webClient;
        this.retry = retryConfig(
                uri.toString(),
                DEFAULT_MAX_ATTEMPTS,
                DEFAULT_INITIAL_WAIT_INTERVAL_MILLIS,
                DEFAULT_EXPONENTIAL_BACKOFF_MULTIPLIER,
                new Class[]{HttpServerErrorException.class},
                logger);
    }

    @Override
    public void scan(String filnavn, byte[] data, String behandlingsId, String fileType) {
        if (enabled && isInfected(filnavn, data, behandlingsId, fileType)) {
            throw new OpplastingException(String.format("Fant virus i fil for behandlingsId %s", behandlingsId), null, "vedlegg.opplasting.feil.muligVirus");
        } else if (!enabled) {
            logger.info("Virusscanning er ikke aktivert");
        }
    }

    private boolean isInfected(String filnavn, byte[] data, String behandlingsId, String fileType) {
        try {
            if (ServiceUtils.isNonProduction() && filnavn.startsWith("virustest")) {
                return true;
            }
            logger.info("Scanner {} bytes for fileType {} (fra Tika)", data.length, fileType);

            ScanResult[] scanResults = withRetry(retry, () -> putForObject(data));

            if (scanResults.length != 1) {
                logger.warn("Uventet respons med lengde {}, forventet lengde er 1", scanResults.length);
                return false;
            }
            ScanResult scanResult = scanResults[0];
            if (OK.equals(scanResult.getResult())) {
                logger.info("Ingen virus i fil");
                return false;
            }
            logger.warn("Fant virus i fil for behandlingsId {}, status {}", behandlingsId, scanResult.getResult());
            return true;
        } catch (Exception e) {
            logger.warn("Kunne ikke scanne fil for behandlingsId {}", behandlingsId, e);
            return false;
        }
    }

    private ScanResult[] putForObject(Object payload) {
        return webClient.put()
                .body(BodyInserters.fromValue(payload))
                .retrieve()
                .bodyToMono(ScanResult[].class)
                .block();
    }
}

File: src/main/kotlin/no/nav/sosialhjelp/soknad/vedlegg/virusscan/ClamAvVirusScanner.kt
package no.nav.sosialhjelp.soknad.vedlegg.virusscan

import kotlinx.coroutines.runBlocking
import no.nav.sosialhjelp.kotlin.utils.retry
import no.nav.sosialhjelp.soknad.client.config.RetryUtils.DEFAULT_EXPONENTIAL_BACKOFF_MULTIPLIER
import no.nav.sosialhjelp.soknad.client.config.RetryUtils.DEFAULT_INITIAL_WAIT_INTERVAL_MILLIS
import no.nav.sosialhjelp.soknad.client.config.RetryUtils.DEFAULT_MAX_ATTEMPTS
import no.nav.sosialhjelp.soknad.common.ServiceUtils
import no.nav.sosialhjelp.soknad.domain.model.exception.OpplastingException
import no.nav.sosialhjelp.soknad.vedlegg.virusscan.dto.Result
import no.nav.sosialhjelp.soknad.vedlegg.virusscan.dto.ScanResult
import org.slf4j.LoggerFactory.getLogger
import org.springframework.web.client.HttpServerErrorException
import org.springframework.web.reactive.function.BodyInserters
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.awaitBody

/**
 * Integrasjonen er kopiert fra https://github.com/navikt/foreldrepengesoknad-api og modifisert til eget bruk
 */
interface VirusScanner {
    fun scan(filnavn: String, data: ByteArray, behandlingsId: String, fileType: String)
}

class ClamAvVirusScanner(
    private val virusScannerWebClient: WebClient,
    private val enabled: Boolean,
    private val serviceUtils: ServiceUtils
) : VirusScanner {

    override fun scan(filnavn: String, data: ByteArray, behandlingsId: String, fileType: String) {
        if (enabled && isInfected(filnavn, data, behandlingsId, fileType)) {
            throw OpplastingException(
                "Fant virus i fil for behandlingsId $behandlingsId",
                null,
                "vedlegg.opplasting.feil.muligVirus"
            )
        } else if (!enabled) {
            log.info("Virusscanning er ikke aktivert")
        }
    }

    private fun isInfected(filnavn: String, data: ByteArray, behandlingsId: String, fileType: String): Boolean {
        try {
            if (serviceUtils.isNonProduction() && filnavn.startsWith("virustest")) {
                return true
            }
            log.info("Scanner ${data.size} bytes for fileType $fileType (fra Tika)")

            val scanResults = runBlocking {
                retry(
                    attempts = DEFAULT_MAX_ATTEMPTS,
                    initialDelay = DEFAULT_INITIAL_WAIT_INTERVAL_MILLIS,
                    factor = DEFAULT_EXPONENTIAL_BACKOFF_MULTIPLIER,
                    retryableExceptions = arrayOf(HttpServerErrorException::class)
                ) {
                    virusScannerWebClient
                        .put()
                        .body(BodyInserters.fromValue(data))
                        .retrieve()
                        .awaitBody<Array<ScanResult>>()
                }
            }

            if (scanResults.size != 1) {
                log.warn("Uventet respons med lengde ${scanResults.size}, forventet lengde er 1")
                return false
            }

            val scanResult = scanResults[0]
            if (Result.OK == scanResult.result) {
                log.info("Ingen virus i fil")
                return false
            }
            log.warn("Fant virus i fil for behandlingsId $behandlingsId, status ${scanResult.result}")
            return true
        } catch (e: Exception) {
            log.warn("Kunne ikke scanne fil for behandlingsId $behandlingsId", e)
            return false
        }
    }

    companion object {
        private val log = getLogger(ClamAvVirusScanner::class.java)
    }
}
